synapse/storage/databases/main/end_to_end_keys.py

```diff
@@ -164,10 +169,36 @@ async def get_e2e_device_keys_for_cs_api(
 
         return rv
 
+    @overload
+    async def get_e2e_device_keys_and_signatures(
+        self,
+        query_list: Collection[Tuple[str, Optional[str]]],
+        include_all_devices: Literal[False] = False,
+    ) -> Dict[str, Dict[str, DeviceKeyLookupResult]]:
+        ...
+
+    @overload
+    async def get_e2e_device_keys_and_signatures(
+        self,
+        query_list: Collection[Tuple[str, Optional[str]]],
+        include_all_devices: bool = False,
+        include_deleted_devices: Literal[False] = False,
+    ) -> Dict[str, Dict[str, DeviceKeyLookupResult]]:
+        ...
+
+    @overload
+    async def get_e2e_device_keys_and_signatures(
+        self,
+        query_list: Collection[Tuple[str, Optional[str]]],
+        include_all_devices: Literal[True],
+        include_deleted_devices: Literal[True],
+    ) -> Dict[str, Dict[str, Optional[DeviceKeyLookupResult]]]:
+        ...
+
     @trace
     async def get_e2e_device_keys_and_signatures(
         self,
-        query_list: List[Tuple[str, Optional[str]]],
+        query_list: Collection[Tuple[str, Optional[str]]],
         include_all_devices: bool = False,
         include_deleted_devices: bool = False,
     ) -> Dict[str, Dict[str, Optional[DeviceKeyLookupResult]]]:
```

Comment:
```
synapse/storage/databases/main/end_to_end_keys.py:198: error: Overloaded function implementation cannot produce return type of signature 1  [misc]
synapse/storage/databases/main/end_to_end_keys.py:198: error: Overloaded function implementation cannot produce return type of signature 2  [misc]
```

I think the complaint is that the return value from the actual function

```
Dict[str, Dict[str, Optional[DeviceKeyLookupResult]]]
```

can never be

```
Dict[str, Dict[str, DeviceKeyLookupResult]]
```

in the first two overloads. I'd guess this is our old friend "Dict is invariant" somehow.

Reply:
Suggestions:
- use `Mapping` or `MutableMapping` in place of `Dict`
- Try making the return type this monstrosity

```
Union[
    Dict[str, Dict[str, DeviceKeyLookupResult]],
    Dict[str, Dict[str, Optional[DeviceKeyLookupResult]]]
]
```

Reply:
> Try making the return type this monstrosity

This works for me locally!

Reply:
I don't really understand -- `DeviceKeyLookupResult` is a subset of `Optional[DeviceKeyLookupResult]` so how can i never be that?

Reply:
> I don't really understand -- `DeviceKeyLookupResult` is a subset of `Optional[DeviceKeyLookupResult]` so how can i never be that?

Dict is invariant in both its parameters, meaning that Dict[K1, V1] is not compatible* with any of the following:

- Dict[K1, V2]
- Dict[K2, V1]
- Dict[K2, V2]

* for some definition of "compatible" which I don't understand well enough to express precisely.

I can't remember why Dict is like this... but it is.  Let me see if I can dig up a rationale link...

(Mapping is covariant in its value type, invariant in its key type.)